### PR TITLE
openstack clone server groups

### PR DESCRIPTION
**Pull request by emjburns**
Following Kubernetes commit stream: https://github.com/lwander/clouddriver/commit/a4772c4dc751f54c91f0da345fada43aabc41cde

Issue: https://github.com/spinnaker/spinnaker/issues/923

### CODE REVIEW

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy [hunk 1]
```diff
@@ -140,6 +140,34 @@ abstract class OpenstackClientProvider {
     //TODO: Handle heat autoscaling migration to senlin in versions > Mitaka
   }
 
+  /***
+   * Get a Spinnaker Server Group (Openstack Heat Stack).
+   * @param stackName
+   */
+  Stack getServerGroup(String stackName) {
+    try {
+      def stack = client.heat().stacks().getStackByName(stackName)
```
Comment (ashleykasim) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy]: should support regions: `client.useRegion()` (apparently a refactor to have a getRegionClient() is forthcoming, but it hasn't been opened a PR for)

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy [hunk 2]
```diff
@@ -140,6 +140,34 @@ abstract class OpenstackClientProvider {
...
+
+  /***
+   * Get a heat template from an existing Openstack Heat Stack
+   * @param stackName
+   * @param stackId
+   * @return
+     */
+  String getHeatTemplate(String stackName, String stackId) {
+    try {
+      def template = client.heat().templates().getTemplateAsString(stackName, stackId)
```
Comment (ashleykasim) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy]: same with this one.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/converters/securitygroup/UpsertOpenstackSecurityGroupAtomicOperationConverter.groovy [hunk 3]
```diff
@@ -26,7 +26,7 @@ import com.netflix.spinnaker.clouddriver.security.AbstractAtomicOperationsCreden
 import org.springframework.stereotype.Component
 
 @OpenstackOperation(AtomicOperations.UPSERT_SECURITY_GROUP)
-@Component("upsertSecurityGroupDescription")
+@Component("upsertOpenstackSecurityGroupDescription")
```
Comment (ashleykasim) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/converters/securitygroup/UpsertOpenstackSecurityGroupAtomicOperationConverter.groovy]: According to cfieber this is legacy naming, we should use just `@Component`

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy (hunk 1 above)
Comment (emjburns) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy]: Fixed.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy (hunk 2 above)
Comment (emjburns) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy]: Fixed.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/converters/securitygroup/UpsertOpenstackSecurityGroupAtomicOperationConverter.groovy (hunk 3 above)
Comment (emjburns) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/converters/securitygroup/UpsertOpenstackSecurityGroupAtomicOperationConverter.groovy]: Removed for all similar ones in this PR. Will leave other refactoring to new PR.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientV2Provider.groovy [hunk 4]
```diff
@@ -43,4 +43,9 @@ class OpenstackClientV2Provider extends OpenstackClientProvider {
 
   //TODO v2 specific operations
 
+  @Override
+  String getRegion() {
+    // v2 does not have a way to access the region
+    return ""
```
Comment (lwander) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientV2Provider.groovy]: A lot of operations/requests are keyed by region when Gate is doing a lookup. This should probably be something like `"NA"` or `"default"`

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/CloneOpenstackAtomicOperation.groovy [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  CloneOpenstackAtomicOperation(CloneOpenstackAtomicOperationDescription description) {
+    this.description = description
+  }
+
+  private static Task getTask() {
+    TaskRepository.threadLocalTask.get()
+  }
+
+  /*
+    * curl -X POST -H "Content-Type: application/json" -d '[{"cloneServerGroup": {"source": {"stackName": "myapp-teststack-v000"},"account": "test"}}]' localhost:7002/openstack/ops
```
Comment (lwander) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/CloneOpenstackAtomicOperation.groovy]: Indentation is off

File: clouddriver-openstack/src/test/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProviderSpec.groovy [hunk 6]
```diff
@@ -64,6 +61,11 @@ class OpenstackClientProviderSpec extends Specification {
       String getTokenId() {
         null
       }
+
+      @Override
+      String getRegion(){
+        return ''
```
Comment (lwander) [inline clouddriver-openstack/src/test/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProviderSpec.groovy]: Try and stay consistent with the quote type used, seems like you use `""` everywhere else

**Comment by lwander** [comment]
LGTM after comments

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientV2Provider.groovy (hunk 4 above)
Comment (emjburns) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientV2Provider.groovy]: Fixing to not need this. Specifying ancestor region in the clone call.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientV3Provider.groovy [hunk 7]
```diff
@@ -43,4 +43,8 @@ class OpenstackClientV3Provider extends OpenstackClientProvider {
 
   //TODO v3 specific operations
 
+  @Override
+  String getRegion() {
+    return client.identity().regions().list().get(0)
```
Comment (emjburns) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientV3Provider.groovy]: Fixing to not need this. Specifying ancestor region in the clone call.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/CloneOpenstackAtomicOperation.groovy (hunk 5 above)
Comment (emjburns) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/CloneOpenstackAtomicOperation.groovy]: Fixing.

**Comment by lwander** [comment]
Will merge once squashed

File: clouddriver-openstack/src/test/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/CloneOpenstackAtomicOperationSpec.groovy [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+  private static final SEQUENCE = "v000"
+  private static final ANCESTOR_STACK_NAME = "$APPLICATION-$STACK-$DETAILS-$SEQUENCE"
+  private static final ANCESTOR_STACK_ID = "978f378f-0b98-469d-90dd-f61a73f8703a"
+
+  // Changed Parameters
+  private static final STACK_N = "stackn"
+  private static final APPLICATION_N = "appn"
+  private static final DETAILS_N = "detailn"
+  private static final REGION_N = "regionn"
+  private static final String HEAT_TEMPLATE_N = "{\"heat_template_version\":\"2013-05-24\"," +
```
Comment (varikin) [inline clouddriver-openstack/src/test/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/CloneOpenstackAtomicOperationSpec.groovy]: If you use single quotes for the entire template, you shouldn't need to escape all the double quotes in the JSON.

**Comment by emjburns** [comment]
Closing and reopening with commit squashed. Screwed up my git :(

**Pull request closed without merging**